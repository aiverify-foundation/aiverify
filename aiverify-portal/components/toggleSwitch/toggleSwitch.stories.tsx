import type { Meta, StoryObj } from '@storybook/react';
import { ToggleSwitch } from './toggleSwitch';
import { userEvent, within, expect } from '@storybook/test';

const meta: Meta<typeof ToggleSwitch> = {
  title: 'Components/Toggle Switch',
  component: ToggleSwitch,
  args: {
    label: 'Toggle Me',
  },
  argTypes: {
    name: {
      table: {
        disable: true,
      },
    },
    value: {
      table: {
        disable: true,
      },
    },
    defaultChecked: {
      table: {
        disable: true,
      },
    },
    onChange: {
      table: {
        disable: true,
      },
    },
  },
};

export default meta;
type Story = StoryObj<typeof ToggleSwitch>;

export const Default: Story = {
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    await userEvent.click(canvas.getByRole('toggle-switch'));
    await expect(canvas.getByRole('checkbox')).toBeChecked();
    await userEvent.click(canvas.getByRole('toggle-switch'));
    await expect(canvas.getByRole('checkbox')).not.toBeChecked();
  },
};
