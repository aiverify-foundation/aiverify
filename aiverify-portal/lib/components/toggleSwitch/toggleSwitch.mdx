import { Meta, Story, Canvas, Primary, Controls, Title } from '@storybook/addon-docs';
import ToggleSwitchStories from './toggleSwitch.stories';

<Title>Toggle Switch React Component</Title>
<p>Toggle Switch allows the user to toggle between two states. Under the hood, it's just a checkbox. So you can bind to it like a normal checkbox.</p>

<Meta of={ToggleSwitchStories} />

<Primary />

<Controls />

In unit tests with RTL, similar to Checkbox the recommended selector is `screen.getByRole('checkbox', { name: /your toggle/i });`
where "your toggle" is the value of the `label` prop.