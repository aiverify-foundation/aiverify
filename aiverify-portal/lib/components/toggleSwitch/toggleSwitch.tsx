'use client';

import { useState } from 'react';
import styles from './styles/toggleSwitch.module.css';

interface ToggleSwitchProps {
  name?: string;
  label?: string;
  value?: string;
  defaultChecked?: boolean;
  onChange?: (isChecked: boolean) => void;
}

function ToggleSwitch(props: ToggleSwitchProps) {
  const { name, label, value, defaultChecked = false, onChange } = props;
  const [isChecked, setIsChecked] = useState(defaultChecked);

  function toggle() {
    const newState = !isChecked;
    setIsChecked(newState);
    if (onChange) {
      onChange(newState);
    }
  }

  return (
    <label className={styles.toggleSwitch}>
      <div role="toggle-switch" className={styles.toggleSwitchContainer} onClick={toggle}>
        <div
          className={`${styles.toggleSwitchBackground} ${isChecked ? styles.toggleSwitchBackgroundChecked : ''}`}
        />
        <div
          className={`${styles.toggleSwitchKnob} ${isChecked ? styles.toggleSwitchKnobChecked : ''}`}
        />
      </div>
      {label && <div className={styles.toggleSwitchLabel}>{label}</div>}
      <input
        readOnly
        type="checkbox"
        defaultChecked={isChecked}
        defaultValue={value}
        name={name}
        style={{ position: 'absolute', opacity: 0, pointerEvents: 'none' }}
      />
    </label>
  );
}

export { ToggleSwitch };
