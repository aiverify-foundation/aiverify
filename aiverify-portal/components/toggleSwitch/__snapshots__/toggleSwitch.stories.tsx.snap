// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Toggle Switch Default play-test 1`] = `
<div style="position: relative; width: 100%; height: 300px; display: flex; align-items: center; justify-content: center;">
  <label class="_toggleSwitch_m4i19_1">
    <div role="toggle-switch"
         class="_toggleSwitchContainer_m4i19_13"
    >
      <div class="_toggleSwitchBackground_m4i19_21 ">
      </div>
      <div class="_toggleSwitchKnob_m4i19_47 ">
      </div>
    </div>
    <div class="_toggleSwitchLabel_m4i19_81">
      Toggle Me
    </div>
    <input readonly
           type="checkbox"
           value
           style="position: absolute; opacity: 0; pointer-events: none;"
    >
  </label>
</div>
`;
